File: apps/backend/app/schemas/memory.py
"""
Pydantic schemas for Coodara V2 Architecture Memory.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ArchitectureComponentResponse(BaseModel):
    """Component entity tracked in architectural memory."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    component_type: str
    path: str | None = None
    description: str | None = None
    status: str
    first_seen_analysis_id: int | None = None
    last_seen_analysis_id: int | None = None
    created_at: datetime
    updated_at: datetime


class ArchitectureRelationshipResponse(BaseModel):
    """Component relationship entity tracked in architectural memory."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source_component_id: int
    target_component_id: int
    kind: str
    status: str
    first_seen_analysis_id: int | None = None
    last_seen_analysis_id: int | None = None
    created_at: datetime
    updated_at: datetime


class ArchitectureTechnologyMemoryResponse(BaseModel):
    """Technology entity tracked in architectural memory across versions."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    technology: str
    version: str | None = None
    category: str | None = None
    status: str
    first_seen_analysis_id: int | None = None
    last_seen_analysis_id: int | None = None
    created_at: datetime
    updated_at: datetime


class ArchitectureMemoryEntryResponse(BaseModel):
    """Discrete architectural knowledge entry with provenance."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    memory_id: int
    organization_id: int
    repository_id: int
    analysis_id: int | None = None
    memory_type: str
    title: str
    content: str
    confidence: float
    source_analyzer: str | None = None
    source_file: str | None = None
    created_at: datetime
    updated_at: datetime


class ArchitectureEventResponse(BaseModel):
    """Historical architectural evolution event."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    organization_id: int
    repository_id: int
    analysis_id: int
    event_type: str
    title: str
    description: str
    details: str | None = None
    created_at: datetime


class ArchitectureMemoryOverviewResponse(BaseModel):
    """Comprehensive architecture memory overview for a repository."""

    repository_id: int
    organization_id: int
    latest_analysis_id: int | None = None
    components_count: int
    active_components_count: int
    technologies_count: int
    entries_count: int
    events_count: int
    components: list[ArchitectureComponentResponse] = Field(default_factory=list)
    technologies: list[ArchitectureTechnologyMemoryResponse] = Field(default_factory=list)
    entries: list[ArchitectureMemoryEntryResponse] = Field(default_factory=list)
    recent_events: list[ArchitectureEventResponse] = Field(default_factory=list)


class ScoredMemoryEntryResponse(BaseModel):
    """Search match for an architectural memory entry."""

    entry: ArchitectureMemoryEntryResponse
    score: float
    matched_terms: list[str] = Field(default_factory=list)


class MemorySearchResponse(BaseModel):
    """Memory search query response."""

    query: str
    total_matches: int
    results: list[ScoredMemoryEntryResponse]


class ArchitectureHistoryResponse(BaseModel):
    """Chronological architectural evolution timeline."""

    repository_id: int
    organization_id: int
    total_events: int
    events: list[ArchitectureEventResponse] = Field(default_factory=list)
    items: list[ArchitectureEventResponse] = Field(default_factory=list)

File: apps/frontend/src/api/memory.ts
/**
 * Architecture Memory API client for Coodara V2.
 */

import { api } from "./client";

export interface ArchitectureComponent {
  id: number;
  name: string;
  component_type: string;
  path?: string | null;
  description?: string | null;
  status: "active" | "removed" | "upgraded";
  first_seen_analysis_id?: number | null;
  last_seen_analysis_id?: number | null;
  created_at: string;
  updated_at: string;
}

export interface ArchitectureTechnologyMemory {
  id: number;
  technology: string;
  version?: string | null;
  category?: string | null;
  status: "active" | "removed" | "upgraded";
  first_seen_analysis_id?: number | null;
  last_seen_analysis_id?: number | null;
  created_at: string;
  updated_at: string;
}

export interface ArchitectureMemoryEntry {
  id: number;
  memory_id: number;
  organization_id: number;
  repository_id: number;
  analysis_id?: number | null;
  memory_type: string;
  title: string;
  content: string;
  confidence: number;
  source_analyzer?: string | null;
  source_file?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ArchitectureEvent {
  id: number;
  organization_id: number;
  repository_id: number;
  analysis_id: number;
  event_type: string;
  title: string;
  description: string;
  details?: string | null;
  created_at: string;
}

export interface ArchitectureMemoryOverview {
  repository_id: number;
  organization_id: number;
  latest_analysis_id?: number | null;
  components_count: number;
  active_components_count: number;
  technologies_count: number;
  entries_count: number;
  events_count: number;
  components: ArchitectureComponent[];
  technologies: ArchitectureTechnologyMemory[];
  entries: ArchitectureMemoryEntry[];
  recent_events: ArchitectureEvent[];
}

export interface ScoredMemoryEntry {
  entry: ArchitectureMemoryEntry;
  score: number;
  matched_terms: string[];
}

export interface MemorySearchResponse {
  query: string;
  total_matches: number;
  results: ScoredMemoryEntry[];
}

export interface ArchitectureHistoryResponse {
  repository_id: number;
  organization_id: number;
  total_events: number;
  events: ArchitectureEvent[];
}

export async function fetchRepositoryMemory(
  orgId: number | string,
  repoId: number | string
): Promise<ArchitectureMemoryOverview> {
  const { data } = await api.get<ArchitectureMemoryOverview>(
    `/organizations/${orgId}/repositories/${repoId}/memory`
  );
  return data;
}

export async function searchRepositoryMemory(
  orgId: number | string,
  repoId: number | string,
  query: string,
  memoryType?: string
): Promise<MemorySearchResponse> {
  const params: Record<string, string> = { q: query };
  if (memoryType) params.memory_type = memoryType;
  const { data } = await api.get<MemorySearchResponse>(
    `/organizations/${orgId}/repositories/${repoId}/memory/search`,
    { params }
  );
  return data;
}

export async function fetchRepositoryEvents(
  orgId: number | string,
  repoId: number | string,
  offset = 0,
  limit = 50,
  eventType?: string
): Promise<ArchitectureEvent[]> {
  const params: Record<string, string | number> = { offset, limit };
  if (eventType) params.event_type = eventType;
  const { data } = await api.get<ArchitectureEvent[]>(
    `/organizations/${orgId}/repositories/${repoId}/memory/events`,
    { params }
  );
  return data;
}

export async function fetchRepositoryHistory(
  orgId: number | string,
  repoId: number | string,
  offset = 0,
  limit = 50
): Promise<ArchitectureHistoryResponse> {
  const { data } = await api.get<ArchitectureHistoryResponse>(
    `/organizations/${orgId}/repositories/${repoId}/architecture/history`,
    { params: { offset, limit } }
  );
  return data;
}
